akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -1112,36 +1116,45 @@ private[cluster] final class JoinSeedNodeProcess(seedNodes: immutable.IndexedSeq
 /**
  * INTERNAL API
  *
- * The supplied callback will be run, once, when current cluster member is `Up`.
+ * The supplied callback will be run, once, when current cluster member come up with the same status.
  */
-private[cluster] class OnMemberUpListener(callback: Runnable) extends Actor with ActorLogging {
+private[cluster] class OnMemberStatusChangedListener(callback: Runnable, status: MemberStatus) extends Actor with ActorLogging {
   import ClusterEvent._
-  val cluster = Cluster(context.system)
-  // subscribe to MemberUp, re-subscribe when restart
+  private val cluster = Cluster(context.system)
+  private val to = status match {
+    case Up ⇒
+      classOf[MemberUp]
+    case Removed ⇒
+      classOf[MemberRemoved]
+  }
+
   override def preStart(): Unit =
-    cluster.subscribe(self, classOf[MemberUp])
+    cluster.subscribe(self, to)
   override def postStop(): Unit =
     cluster.unsubscribe(self)
 
   def receive = {
     case state: CurrentClusterState ⇒
-      if (state.members.exists(isSelfUp(_)))
+      if (state.members.exists(isTriggered))
+        done()
+    case MemberUp(member) ⇒
+      if (isTriggered(member))
         done()
-    case MemberUp(m) ⇒
-      if (isSelfUp(m))
+    case MemberRemoved(member, _) ⇒
+      if (isTriggered(member))
         done()
   }
 
-  def done(): Unit = {
+  private def done(): Unit = {
     try callback.run() catch {
```

Comment:
I think the code should not be run in the actor or even the actor system's dispatcher if we want to get around of this ,or using a deadlock detector when the system is shut downing.or we could add a note to the the doc about the```DO NOT BLOCK```.